### DISCUSSION TITLE
The trait MultiConnection is not implemented

### DISCUSSION THREAD

**Discussion by Ludea** [weiznich/diesel_async#108] [Ideas]
### Versions

- **Rust:** 1.72.0
- **Diesel:** 2.1.1
- **Diesel_async:** 0.4.1
- **Database:** Mysql / Postgresql
- **Operating System** Ubuntu 22.04

### Feature Flags

- **diesel:** 
- **diesel_async:** mysql, postresql

## Problem Description
I can't use MultiConnection with async Connection

### What are you trying to accomplish?
I would like to use MultiConnection trait with async connection

### What is the expected output?
No issue

### What is the actual output?
`the trait diesel::Connection is not implemented for diesel_async::AsyncMysqlConnection`

### Steps to reproduce
```
#[derive(diesel::MultiConnection)]
pub enum InferConnection {
    Pg(AsyncPgConnection),
    Mysql(AsyncMysqlConnection),
}
```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/weiznich/diesel_async/issues) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
This is expected behavior as the `#[derive(MultiConnection)]` proc macro is designed to be used with `diesel::Connection` types and not with async connection types. It's not a focus for me to provide this functionality for the async crate as well. If someone is interested in this I'm happy to review a PR that includes tests + documentation for this. Otherwise it's also possible to implement that outside of diesel-async.

**Reply by avitex** [reply]
Hello @weiznich. Thank you for your work on diesel-async! I tried investigating whether I could implement a multi-connection outside of `diesel-async`, and ran into a few issues that I couldn't work past.

1. Send Bounds
Example in the source here: https://gist.github.com/avitex/bf43406dc682d74c6cab04953a86f03c#file-manual-multi-connection-rs-L45-L57, `SerializedQuery` doesn't implement `Send` which is required to be returned in the `LoadFuture` (which has a `Send` bound). `SerializedQuery` itself can't implement `Send` due to the `inner` field (initialized with `source.as_query()`), holding a type implementing the `Query` trait which is not `Send`. This might be solved with an additional `Send` bound on the `AsyncConnectionCore::load` function.

2. The `AnsiTransactionManager` from `diesel-async` and `diesel` are different, and `SyncConnectionWrapper` uses the `AnsiTransactionManager` from `diesel` for sqlite, which conflicts here https://gist.github.com/avitex/bf43406dc682d74c6cab04953a86f03c#file-manual-multi-connection-rs-L132-L133. This may be avoidable with an enum wrapper around both different `AnsiTransactionManager` types.

3. Some other issues with lifetimes that I haven't looked deeply into. 

I'm happy to investigate at some point what changes may be required to support implementing a multi-connection, though I wanted to check whether you have any thoughts or suggestions at this point? If you think it's worth the time, I'll also setup a repository with the code I referenced for reproducing what I've tried so far.

**Reply by weiznich** [reply]
Thanks for looking into this. That's already helpful :heart:

> Send Bounds
Example in the source here: https://gist.github.com/avitex/bf43406dc682d74c6cab04953a86f03c#file-manual-multi-connection-rs-L45-L57, SerializedQuery doesn't implement Send which is required to be returned in the LoadFuture (which has a Send bound). SerializedQuery itself can't implement Send due to the inner field (initialized with source.as_query()), holding a type implementing the Query trait which is not Send. This might be solved with an additional Send bound on the AsyncConnectionCore::load function.

I think it should be fine to just add a send point to `AsyncConnectionCore::load` for the generic type there.

> The AnsiTransactionManager from diesel-async and diesel are different, and SyncConnectionWrapper uses the AnsiTransactionManager from diesel for sqlite, which conflicts here https://gist.github.com/avitex/bf43406dc682d74c6cab04953a86f03c#file-manual-multi-connection-rs-L132-L133. This may be avoidable with an enum wrapper around both different AnsiTransactionManager types.

The sync implementation just implements a custom `TransactionManager` version for the `MultiConnection` enum. It essentially just matches on the different variants there. See for the actual generated code

* https://github.com/diesel-rs/diesel/blob/main/diesel_derives/src/tests/snapshots/diesel_derives__tests__multiconnection_1.snap#L1774
* https://github.com/diesel-rs/diesel/blob/main/diesel_derives/src/tests/snapshots/diesel_derives__tests__multiconnection_1.snap#L1932-L2010

The same pattern should work for the async implementation (with different traits) as well.

> Some other issues with lifetimes that I haven't looked deeply into.

Having more details on this would be helpful to tell whether that's a problem or not.

More generally speaking: For me the biggest blocker is figuring out where a good way to support that in the `#[derive(MultiConnection)]` implementation. That one lives in `diesel_derives` that has currently no knowledge about `diesel-async`.

**Reply by jusax23** [reply]
Hi, Thanks for the work on diesel-async. (If there is any work going on about this discussion, im not aware of, please let me now.)
I used @avitex code and got it to work with a few modifications (except with postgres).
I did not do extensive testing but a simple test with migrations, insert and select works.

1. ~~I've added the Send requirement in a local copy. To still get it to compile i had to add it in mulitple places. Including 4 places in the main `diesel` repo (Send and/or Sync).~~ (edit: Seems to work without adding the Send, the way its in the gist below. Moving the line `let loading = conn.load(query);` out of the async, boxed block made that change unnecessary.)

2. I implemented this like suggested from @weiznich (and prepared in the comment below avitex's version)

3. I stumbeld on the lifetimes to. https://gist.github.com/jusax23/c3c342adf90350e06a63843fb7f05636#file-manual-multi-connection-rs-L58-L65
With some modification outside `diesel-async` and a change in the `SyncConnectionWrapper`:
```diff
// src/sync_connection_wrapper/mod.rs
@@ -152,9 +152,8 @@ mod implementation {
         // SpawnBlocking bounds
         S: SpawnBlocking + Send,
     {
-        type LoadFuture<'conn, 'query> =
-            BoxFuture<'query, QueryResult<Self::Stream<'conn, 'query>>>;
-        type ExecuteFuture<'conn, 'query> = BoxFuture<'query, QueryResult<usize>>;
+        type LoadFuture<'conn, 'query> = BoxFuture<'conn, QueryResult<Self::Stream<'conn, 'query>>>;
+        type ExecuteFuture<'conn, 'query> = BoxFuture<'conn, QueryResult<usize>>;
         type Stream<'conn, 'query> = BoxStream<'static, QueryResult<Self::Row<'conn, 'query>>>;
         type Row<'conn, 'query> = O;
         type Backend = <C as Connection>::Backend;
```
I got it to compile. Is the change from `'query` to `'conn` feasible? Like this it matches more closely with mysql and pg. And i can hold `&'conn mut self` (the AnyConn object) as long/long enough in the return type.

The Problem with postgres: https://gist.github.com/jusax23/c3c342adf90350e06a63843fb7f05636#file-manual-multi-connection-rs-L282-L308
I tried to fix the `MetadataLookup` by implementing a copy of `MultiConnectionHelper`.
This works well for Mysql and Sqlite. 
I added the `as_any` implementation of `impl<F> PgMetadataLookup for PgAsyncMetadataLookup<F>` to diesel-async.
But i can't figure out the type that should be used for the downcast with pg. (I think the `F` generic could be a part of the problem?)

In my opinion a good place for this would be a new `#[derive(AsyncMultiConnection)]` macro. (Im not sure where the right place for this would be and it will probable duplicate some code, thats the same between the async and sync version.)

ps. I'm not sure about copyright, because its based on the sync implementation. But @avitex if it is not okay if i use/publish your code/my modification please let me now?

**Reply by jusax23** [reply]
Hi,
I found a working solution for postgres, by removeing the mentioned `F` generic and using a Box.
I put all changes, i think are nessecary (or in the case of `AsyncMultiConnectionHelper` useful),  in this branch: https://github.com/jusax23/diesel_async/tree/dev-async_multi_connection

@weiznich i would be willing to open a pr, if you think these changes are okay.

I managed to port the `#[derive(MultiConnection)]` to a async version `#[derive(MultiConnection)]` (with 94 insertions(+), 165 deletions(-)).

**Reply by weiznich** [reply]
> I got it to compile. Is the change from 'query to 'conn feasible? Like this it matches more closely with mysql and pg. And i can hold &'conn mut self (the AnyConn object) as long/long enough in the return type.


I think this is acceptable.

> @weiznich i would be willing to open a pr, if you think these changes are okay.

Please do, that would be very welcome

> I managed to port the #[derive(MultiConnection)] to a async version #[derive(MultiConnection)] (with 94 insertions(+), 165 deletions(-)).

My suggestion would be to add a `#[diesel_async]` attribute to the existing `#[derive(MultiConnection)]` proc-macro that then conditionally generates code for diesel-async instead of diesel. That lets us share most of the code.

We then would also need to have an async copy of most of the stuff in https://github.com/diesel-rs/diesel/blob/main/diesel_derives/tests/multiconnection.rs in diesel-async.

Additionally the new attribute needs to be documented and somewhat tested in diesel_derives. (Most likely adding a test [here](https://github.com/diesel-rs/diesel/blob/main/diesel_derives/src/tests/multiconnection.rs) is sufficient, that doesn't require any extra dependencies)
